In [1]:
# installation of required libraries and dependencies
# numeric calculations
! pip install numpy==1.26.0 
# data frames 
! pip install pandas==2.1.1 
# machine learning algorithms 
! pip install scikit-learn==1.5.1 
! pip install scipy==1.12.0
# plotting 
! pip install plotly==5.24.1 
! pip install matplotlib==3.8.0 
! pip install seaborn==0.13.2 
! pip install plotly-express==0.4.1 
! pip install chart-studio==1.1.0 
# web app library 
! pip install streamlit==1.37.1 
# association rules
! pip install mlxtend==0.23.3

! pip install nbformat==5.9.2 

Defaulting to user installation because normal site-packages is not writeable


ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11; 1.26.0 Requires-Python >=3.9,<3.13; 1.26.1 Requires-Python >=3.9,<3.13
ERROR: Could not find a version that satisfies the requirement numpy==1.26.0 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.22

Defaulting to user installation because normal site-packages is not writeable
  Using cached pandas-2.1.1.tar.gz (4.3 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'


  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [12 lines of output]
      + meson setup C:\Users\tomek\AppData\Local\Temp\pip-install-pe0r0t2u\pandas_e326df1f0e604066b2271f334adab2b3 C:\Users\tomek\AppData\Local\Temp\pip-install-pe0r0t2u\pandas_e326df1f0e604066b2271f334adab2b3\.mesonpy-b28hnlho\build -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --vsenv --native-file=C:\Users\tomek\AppData\Local\Temp\pip-install-pe0r0t2u\pandas_e326df1f0e604066b2271f334adab2b3\.mesonpy-b28hnlho\build\meson-python-native-file.ini
      The Meson build system
      Version: 1.2.1
      Source dir: C:\Users\tomek\AppData\Local\Temp\pip-install-pe0r0t2u\pandas_e326df1f0e604066b2271f334adab2b3
      Build dir: C:\Users\tomek\AppData\Local\Temp\pip-install-pe0r0t2u\pandas_e326df1f0e604066b2271f334adab2b3\.mesonpy-b28hnlho\build
      Build type: native build
      Project name: pandas
      Project version: 2.1.1
    

Defaulting to user installation because normal site-packages is not writeable
  Using cached scikit_learn-1.5.1.tar.gz (7.0 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'


  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [121 lines of output]
      + meson setup C:\Users\tomek\AppData\Local\Temp\pip-install-xvtg6htt\scikit-learn_016bb58d3abb491baff1b68222235595 C:\Users\tomek\AppData\Local\Temp\pip-install-xvtg6htt\scikit-learn_016bb58d3abb491baff1b68222235595\.mesonpy-u43vva6f -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=C:\Users\tomek\AppData\Local\Temp\pip-install-xvtg6htt\scikit-learn_016bb58d3abb491baff1b68222235595\.mesonpy-u43vva6f\meson-python-native-file.ini
      The Meson build system
      Version: 1.10.0
      Source dir: C:\Users\tomek\AppData\Local\Temp\pip-install-xvtg6htt\scikit-learn_016bb58d3abb491baff1b68222235595
      Build dir: C:\Users\tomek\AppData\Local\Temp\pip-install-xvtg6htt\scikit-learn_016bb58d3abb491baff1b68222235595\.mesonpy-u43vva6f
      Build type: native build
      Project name: scikit-learn
      Project version

Defaulting to user installation because normal site-packages is not writeable
  Using cached scipy-1.12.0.tar.gz (56.8 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'


  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [44 lines of output]
      + meson setup C:\Users\tomek\AppData\Local\Temp\pip-install-vdwk9fx4\scipy_3c541bd7d6ea44f2ac96277125cd3b5f C:\Users\tomek\AppData\Local\Temp\pip-install-vdwk9fx4\scipy_3c541bd7d6ea44f2ac96277125cd3b5f\.mesonpy-ak7vx5d7 -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=C:\Users\tomek\AppData\Local\Temp\pip-install-vdwk9fx4\scipy_3c541bd7d6ea44f2ac96277125cd3b5f\.mesonpy-ak7vx5d7\meson-python-native-file.ini
      The Meson build system
      Version: 1.10.0
      Source dir: C:\Users\tomek\AppData\Local\Temp\pip-install-vdwk9fx4\scipy_3c541bd7d6ea44f2ac96277125cd3b5f
      Build dir: C:\Users\tomek\AppData\Local\Temp\pip-install-vdwk9fx4\scipy_3c541bd7d6ea44f2ac96277125cd3b5f\.mesonpy-ak7vx5d7
      Build type: native build
      Project name: scipy
      Project version: 1.12.0
      C compiler for the host mach

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable
  Using cached matplotlib-3.8.0.tar.gz (35.9 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Using cached numpy-1.26.4-cp313-cp313-win_amd64.whl
Failed to build matplotlib


  error: subprocess-exited-with-error
  
  × Building wheel for matplotlib (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [652 lines of output]
      
      Edit mplsetup.cfg to change the build options; suppress output with --quiet.
      
      BUILDING MATPLOTLIB
            python: yes [3.13.7 (tags/v3.13.7:bcee1c3, Aug 14 2025, 14:15:11) [MSC
                        v.1944 64 bit (AMD64)]]
          platform: yes [win32]
             tests: no  [skipping due to configuration]
            macosx: no  [Mac OS-X only]
      
      toml section missing WindowsPath('pyproject.toml') does not contain a tool.setuptools_scm section
      C:\Users\tomek\AppData\Local\Temp\pip-build-env-29eq4c8_\overlay\Lib\site-packages\setuptools_scm\git.py:427: UserWarning: git archive did not support describe output
        warnings.warn("git archive did not support describe output")
      C:\Users\tomek\AppData\Local\Temp\pip-build-env-29eq4c8_\overlay\Lib\site-packages\setuptools_sc

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
# load pandas to deal with the data
import pandas as pd
# plotting
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# load data from table file where entries are separated with a space
data = pd.read_csv("../data/flickr_data2.csv", sep=",")
# show first 5 rows of the dataframe
data.head()

C:\Users\tomek\AppData\Local\Temp\ipykernel_9540\2003460992.py:2: DtypeWarning: Columns (11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("../data/flickr_data2.csv", sep=",")


id          user        lat      long  \
0  4395181099  30624617@N03  45.754858  4.821710   
1  4394748717  35853470@N00  45.753270  4.862953   
2  4394694699  11817998@N05  45.760655  4.846564   
3  4394803790  11545749@N06  45.784000  4.874072   
4  4394803554  11545749@N06  45.784000  4.874072   

                                                tags  \
0                 chair,lyon,rhône,chaise,rhônealpes   
1                                                NaN   
2                                         365,iphone   
3       nin,nineinchnails,gift,screening,toiou,avott   
4  lyon,nin,nineinchnails,gift,screening,toiou,avott   

                         title   date_taken_minute   date_taken_hour  \
0             Chaises avec vue                11.0                15   
1                          NaN                51.0                17   
2       59/365 - R46 V103 B163                29.0                17   
3  2010-01-29 Toiou Avott Lyon                15.0                20   
4  2010-01-28 Toiou Avott Lyon                10.0                20   

    date_taken_day   date_taken_month   date_taken_year  date_upload_minute  \
0               28                  2              2010                  23   
1               28                  2              2010                  52   
2               28                  2              2010                  33   
3               28                  1              2010                  38   
4               28                  1              2010                  38   

   date_upload_hour   date_upload_day   date_upload_month   date_upload_year  \
0                20              28.0                   2             2010.0   
1                17              28.0                   2             2010.0   
2                17              28.0                   2             2010.0   
3                12              28.0                   2             2010.0   
4                12              28.0                   2             2010.0   

   Unnamed: 16  Unnamed: 17  Unnamed: 18  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN

## Introduction

Analyse 1: Les attributs important semblent être lat, long et l'heure pour regrouper les personnes.

On remarque que les 3 derniéres colonnes sont remplies de NaN, nous allons explorer les détails de ces colonnes 
Les types de "date" et "minute" sont float et non int.

In [4]:
print(len(data))

420240


Il y a 420240 lignes.

## Nettoyage des données 

On évalue et nettoie les doublons 

In [5]:
print(len(data['id'].unique()))

168097


In [6]:
ligne_id_egales = len(data) - len(data['id'].unique())
ligne_completement_egales = len(data) - data.drop_duplicates().shape[0]

print(ligne_id_egales, ligne_completement_egales)


252143 232642


Il y a des id et des lignes dupliquées. 

On ne veut pas supprimer tous les doublons d'id car un doublon pourrait avoir plus d'information qu'un autre (ligne avec des attributs vides)

Comme attendu, ligne_completement_egales< ligne_id_egales.
Nous allons donc dans un premier temps supprimer les lignes entiérement dupliquées.

In [7]:
data_cleaned_1 = data.drop_duplicates() #DataFrame sans doublons sur les lignes

On cherche à savoir si les lignes avec des id identiques sont différentes ou seulement trouées.


In [8]:
duplicated_ids = data_cleaned_1.loc[data_cleaned_1['id'].duplicated(), 'id'].unique().tolist()

data_cleaned_1[data_cleaned_1['id'].isin(duplicated_ids)].to_csv('../temp/duplicated_ids.csv')

On voit que la plupart des doublons sont liés au données type str: Tags, Description. Les données importantes semblent identiques.

Nous allons supprimer les colonnes parasites + les 3 dernières colonnes pour observer les problèmes. Nous traîterons les conflits de texte liés à ces colonnes dans un deuxiéme temps.

In [9]:
cols = ["tags", "title", "Unnamed: 16", "Unnamed: 17", "Unnamed: 18"]
cols_to_drop = [c for c in cols if c in data_cleaned_1.columns]
print(cols, cols_to_drop)
print(data_cleaned_1.columns)

['tags', 'title', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18'] ['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']
Index(['id', ' user', ' lat', ' long', ' tags', ' title', ' date_taken_minute',
       ' date_taken_hour', ' date_taken_day', ' date_taken_month',
       ' date_taken_year', ' date_upload_minute', ' date_upload_hour',
       ' date_upload_day', ' date_upload_month', ' date_upload_year',
       'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18'],
      dtype='object')


On remarque qu'il y avait des espaces dans les colonnes tags et title. Nous allons corriger ça et afficher les mêmes données sans cols.

In [11]:
data_cleaned_2 = data_cleaned_1.copy()
data_cleaned_2.columns = data_cleaned_2.columns.astype(str).str.replace('\xa0', ' ', regex=False).str.lstrip()
cols = data_cleaned_2.columns.tolist()
print(cols)

['id', 'user', 'lat', 'long', 'tags', 'title', 'date_taken_minute', 'date_taken_hour', 'date_taken_day', 'date_taken_month', 'date_taken_year', 'date_upload_minute', 'date_upload_hour', 'date_upload_day', 'date_upload_month', 'date_upload_year', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']


Nous pouvons continuer nos tests.

In [12]:
cols_to_drop = ["tags", "title", "Unnamed: 16", "Unnamed: 17", "Unnamed: 18"]
data_parasites_dropped = data_cleaned_2.drop(columns=cols_to_drop)

data_parasites_dropped[data_parasites_dropped['id'].isin(duplicated_ids)].to_csv('../temp/data_parasites_dropped.csv')

Nous observons des problèmes d'intégrité de la donnée: des mois > 12 ou des valeurs 1 et 1.0 qui sont considérées différentes.

Nous allons dans un premier temps cast les données dans le type adéquat.

In [14]:
data_cleaned_2['id'] = pd.to_numeric(data_cleaned_2['id'], errors='coerce').astype('Int64')

for c in ['lat', 'long']:
    data_cleaned_2[c] = pd.to_numeric(data_cleaned_2[c], errors='coerce').astype(float)

date_cols = [c for c in data_cleaned_2.columns if c.startswith('date_')]
for c in date_cols:
    data_cleaned_2[c] = pd.to_numeric(data_cleaned_2[c], errors='coerce').astype('Int64')

data_cleaned_2[[ 'id', 'lat', 'long' ] + date_cols].info()
data_cleaned_2[[ 'id', 'lat', 'long' ] + date_cols].head()


<class 'pandas.core.frame.DataFrame'>
Index: 187598 entries, 0 to 420239
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  187598 non-null  Int64  
 1   lat                 187598 non-null  float64
 2   long                187598 non-null  float64
 3   date_taken_minute   187597 non-null  Int64  
 4   date_taken_hour     187598 non-null  Int64  
 5   date_taken_day      187598 non-null  Int64  
 6   date_taken_month    187598 non-null  Int64  
 7   date_taken_year     187598 non-null  Int64  
 8   date_upload_minute  187546 non-null  Int64  
 9   date_upload_hour    187596 non-null  Int64  
 10  date_upload_day     187597 non-null  Int64  
 11  date_upload_month   187598 non-null  Int64  
 12  date_upload_year    187597 non-null  Int64  
dtypes: Int64(11), float64(2)
memory usage: 22.0 MB


id        lat      long  date_taken_minute  date_taken_hour  \
0  4395181099  45.754858  4.821710                 11               15   
1  4394748717  45.753270  4.862953                 51               17   
2  4394694699  45.760655  4.846564                 29               17   
3  4394803790  45.784000  4.874072                 15               20   
4  4394803554  45.784000  4.874072                 10               20   

   date_taken_day  date_taken_month  date_taken_year  date_upload_minute  \
0              28                 2             2010                  23   
1              28                 2             2010                  52   
2              28                 2             2010                  33   
3              28                 1             2010                  38   
4              28                 1             2010                  38   

   date_upload_hour  date_upload_day  date_upload_month  date_upload_year  
0                20               28                  2              2010  
1                17               28                  2              2010  
2                17               28                  2              2010  
3                12               28                  2              2010  
4                12               28                  2              2010

Puis supprimer les lignes qui ont des données incohérentes (ex: mois = 120) et supprimer les potentiels nouveaux doublons. On va en profiter pour retirer doublons d'id en gardant la ligne avec le plus d'informations non-nulles

In [20]:
df = data_cleaned_2.copy()

# build validity mask
valid = pd.Series(True, index=df.index)

for col in date_cols:
    if col not in df.columns:
        continue
    if 'month' in col:
        cond = df[col].isna() | ((df[col] >= 1) & (df[col] <= 12))
    elif 'day' in col:
        cond = df[col].isna() | ((df[col] >= 1) & (df[col] <= 31))
    elif 'hour' in col:
        cond = df[col].isna() | ((df[col] >= 0) & (df[col] <= 23))
    elif 'minute' in col:
        cond = df[col].isna() | ((df[col] >= 0) & (df[col] <= 59))
    elif 'year' in col:
        cond = df[col].isna() | ((df[col] >= 1900) & (df[col] <= 2025))
    else:
        cond = df[col].isna()
    valid &= cond
    

if 'lat' in df.columns and 'long' in df.columns:
    valid &= df['lat'].between(-90, 90) & df['long'].between(-180, 180)

print("Lignes totales:", len(df))
print("Valides selon règles date/heure + géo:", valid.sum())
print("Invalides:", (~valid).sum())

df_valid = df.loc[valid].copy()


# retirer doublons d'id en gardant la ligne avec le plus d'informations non-nulles
if 'id' in df_valid.columns:
    df_valid['nn_count'] = df_valid.notna().sum(axis=1)
    df_valid = df_valid.sort_values(['id', 'nn_count'], ascending=[True, False])
    df_dedup = df_valid.drop_duplicates(subset='id', keep='first').drop(columns=['nn_count'])
else:
    df_dedup = df_valid.copy()

print("Après filtration:", len(df_dedup), "lignes")
df_dedup.reset_index(drop=True, inplace=True)
df_dedup.to_csv('../temp/data_cleaned_3.csv', index=False)

data_cleaned_3 = df_dedup


Lignes totales: 187598
Valides selon règles date/heure + géo: 187543
Invalides: 55
Après filtration: 168050 lignes


Maintenant, pour plus de lisibilité, on va concatenate les dates pour avoir moins de colonnes.

In [22]:
# Concatenate date_taken columns into a single datetime column
data_cleaned_4 = data_cleaned_3.copy()

data_cleaned_4['date_taken'] = pd.to_datetime(
    data_cleaned_4[['date_taken_year', 'date_taken_month', 'date_taken_day', 'date_taken_hour', 'date_taken_minute']].astype(str).agg('-'.join, axis=1),
    format='%Y-%m-%d-%H-%M',
    errors='coerce'
)

# Concatenate date_upload columns into a single datetime column
data_cleaned_4['date_upload'] = pd.to_datetime(
    data_cleaned_4[['date_upload_year', 'date_upload_month', 'date_upload_day',  'date_upload_hour', 'date_upload_minute']].astype(str).agg('-'.join, axis=1),
    format='%Y-%m-%d-%H-%M',
    errors='coerce'
)

# Drop the individual date columns
date_cols_to_drop = ['date_taken_minute', 'date_taken_hour', 'date_taken_day', 'date_taken_month', 'date_taken_year', 'date_upload_minute', 'date_upload_hour', 'date_upload_day', 'date_upload_month', 'date_upload_year']
data_cleaned_4 = data_cleaned_4.drop(columns=date_cols_to_drop)

print(data_cleaned_4.head())
print(data_cleaned_4.info())

data_cleaned_4.to_csv('../temp/data_cleaned_4.csv', index=False)

           id          user        lat      long  \
0   306667535  46629827@N00  45.777056  4.857813   
1   306668760  46629827@N00  45.776518  4.857459   
2  1140570280  46629827@N00  45.777681  4.854626   
3  3836705013  46629827@N00  45.797000  4.833338   
4  4076280298  13339562@N06  45.767822  4.836280   

                                                tags  \
0                            france,zoo,lyon,animaux   
1                                    france,zoo,lyon   
2         summer,france,lyon,69,2007,parcdelatêtedor   
3  summer,france,lyon,69,2009,rhônealpes,caluiree...   
4                    opera,lyon,danse,hip,hop,sanggi   

                                               title  Unnamed: 16  \
0                                    P'tits Girafons          NaN   
1                                     Envoyez... :))          NaN   
2                               Parc de la Tête d'Or          NaN   
3  Église romane Notre-Dame, Île Barbe (île situé...          NaN   
4    

Vérifions maintenant s'il reste des doublons, et profitons-en pour regarde ce qu'il y a dans les colonnes unamed.

In [24]:
print("Unnamed: 16 non-null values:")
print(data_cleaned_4[data_cleaned_4['Unnamed: 16'].notna()][['id', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']])

print("\nUnnamed: 17 non-null values:")
print(data_cleaned_4[data_cleaned_4['Unnamed: 17'].notna()][['id', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']])

print("\nUnnamed: 18 non-null values:")
print(data_cleaned_4[data_cleaned_4['Unnamed: 18'].notna()][['id', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']])

print("\nReste des lignes dupliquées:", data_cleaned_4.duplicated().sum())
print("REste des id dupliqués:", len(data_cleaned_4) - len(data_cleaned_4['id'].unique()))

Unnamed: 16 non-null values:
Empty DataFrame
Columns: [id, Unnamed: 16, Unnamed: 17, Unnamed: 18]
Index: []

Unnamed: 17 non-null values:
Empty DataFrame
Columns: [id, Unnamed: 16, Unnamed: 17, Unnamed: 18]
Index: []

Unnamed: 18 non-null values:
Empty DataFrame
Columns: [id, Unnamed: 16, Unnamed: 17, Unnamed: 18]
Index: []

Reste des lignes dupliquées: 0
REste des id dupliqués: 0


Nous avons bien éradiqué les doublons. Avec ce ménage, on remarque que les colones Unamed sont vides, donc on peut les supprimer et passer à la suite.

In [ ]:
data_cleaned_5 = data_cleaned_4.copy()

data_cleaned_5 = data_cleaned_5.drop(columns=["Unnamed: 16", "Unnamed: 17", "Unnamed: 18"])
print(data_cleaned_5.columns)

Index(['id', 'user', 'lat', 'long', 'tags', 'title', 'date_taken',
       'date_upload'],
      dtype='object')


Essayons de tester un dernier cas, ou la date d'upload serait inférieure à la date taken

In [ ]:
data_cleaned_5['time_diff'] = data_cleaned_5['date_upload'] - data_cleaned_5['date_taken']

invalid_dates = data_cleaned_5['time_diff'] < pd.Timedelta(0)
print(f"Lignes avec upload avant taken: {invalid_dates.sum()}")

data_cleaned_5 = data_cleaned_5[~invalid_dates].drop(columns=['time_diff'])

print(f"Le reste des lignes: {len(data_cleaned_5)}")
data_cleaned_5.to_csv('../data/data_cleaned_5.csv', index=False)

Lignes avec upload avant taken: 1925
Le reste des lignes: 166125


## Partie 2 : Affichage de la map et les points

In [32]:
data = pd.read_csv("../data/data_cleaned_5.csv", sep=",")
data.head()

id          user        lat      long  \
0   306667535  46629827@N00  45.777056  4.857813   
1   306668760  46629827@N00  45.776518  4.857459   
2  1140570280  46629827@N00  45.777681  4.854626   
3  3836705013  46629827@N00  45.797000  4.833338   
4  4076280298  13339562@N06  45.767822  4.836280   

                                                tags  \
0                            france,zoo,lyon,animaux   
1                                    france,zoo,lyon   
2         summer,france,lyon,69,2007,parcdelatêtedor   
3  summer,france,lyon,69,2009,rhônealpes,caluiree...   
4                    opera,lyon,danse,hip,hop,sanggi   

                                               title           date_taken  \
0                                    P'tits Girafons  2005-11-01 15:48:00   
1                                     Envoyez... :))  2005-11-01 15:45:00   
2                               Parc de la Tête d'Or  2007-08-14 16:03:00   
3  Église romane Notre-Dame, Île Barbe (île situé...  2009-08-19 15:00:00   
4                                            Hip Hop  2009-11-04 17:07:00   

           date_upload  
0  2014-10-19 12:00:00  
1  2014-07-25 00:00:00  
2  2014-07-17 00:00:00  
3  2018-12-18 00:00:00  
4  2009-11-04 22:42:00

In [ ]:
pip install folium

In [34]:
import folium
from folium.plugins import HeatMap, MarkerCluster

Création d'une hitmap pour une premiére visualisation

In [39]:
df = pd.read_csv("../data/data_cleaned_5.csv", sep=",")

df = df.dropna(subset=['lat', 'long'])

center_lat = df['lat'].mean()
center_lon = df['long'].mean()

map_lyon = folium.Map(
    location=[center_lat, center_lon],
    zoom_start=12,
    tiles='OpenStreetMap'
)


heat_data = [[row['lat'], row['long']] for _, row in df.iterrows()]
HeatMap(heat_data, radius=15, blur=25, max_zoom=13).add_to(map_lyon)

map_lyon.save('../output/carte_lyon_heatmap.html')

Clustering:

In [44]:
from sklearn.cluster import DBSCAN

map_lyon_cluster = folium.Map(
    location=[center_lat, center_lon],
    zoom_start=12,
    tiles='OpenStreetMap'
)


coords = df[['lat', 'long']].values

db = DBSCAN(eps=0.005, min_samples=10).fit(coords)
df['cluster'] = db.labels_

colors = ['red', 'blue', 'green', 'purple', 'orange', 'darkred', 
          'lightred', 'beige', 'darkblue', 'darkgreen', 'cadetblue', 
          'darkpurple', 'pink', 'lightblue', 'lightgreen', 'gray', 
          'black', 'lightgray']

for cluster_id in df['cluster'].unique():
    if cluster_id == -1: 
        continue
    
    cluster_data = df[df['cluster'] == cluster_id]
    color = colors[cluster_id % len(colors)]
    
    for _, row in cluster_data.iterrows():
        folium.CircleMarker(
            location=[row['lat'], row['long']],
            radius=1,
            color=color,
            fill=True,
            fillColor=color,
            fillOpacity=0.7,
            popup=f"Cluster {cluster_id}"
        ).add_to(map_lyon_cluster)

In [46]:
map_lyon_cluster.save('../output/carte_lyon_clustering.html')

On remarque que l'algo fonctionne globalement bien mais est peu efficace sur les zones touristiques (lyon car ne les différencie pas). On ne peut pas augmenter le nombre de cluster max car trop couteux.

Nous allons donc zoomer sur la ville de lyon en choisisant le cluster de Lyon.